File: my-solutions/0230-kth-smallest-element-in-a-bst/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int i = 0;
    int inorder(TreeNode* root, int k) {
        if(root ==  nullptr) {
            return -1;
        }
        if(root->left) {
            int temp = inorder(root->left, k);
            if(temp != -1)
                return temp;
        }
        i++;
        if(i == k)
            return root->val;

        if(root->right) {
            int temp = inorder(root->right, k);
            if(temp != -1)
                return temp;
        }
        
        return -1;
        
    }
    int kthSmallest(TreeNode* root, int k) {
        return inorder(root, k);
    }
};

File: my-solutions/0230-kth-smallest-element-in-a-bst/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        # iterative solution
        inorder = []
        def io(node):
            if not node:
                return
            nonlocal inorder
            io(node.left)
            if len(inorder) == k:
                return
            inorder.append(node.val)
            io(node.right)

        io(root)

        return inorder[k - 1]
